### PR TITLE
first version of solution

### PR DESCRIPTION
**Pull request by Sashk52**

### CODE REVIEW

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    @PostMapping
+    public AccountResponseDto create(@RequestBody AccountRequestDto accountRequestDto) {
+        Account account = accountMapper.toEntity(accountRequestDto);
+        account.setActive(true);
+        return accountMapper.toDto(accountService.save(account));
+    }
+
+    @GetMapping("/by-phone")
+    public AccountResponseDto getAllByPhoneNumber(@RequestParam String accountNumber) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/AccountController.java]: let's rewrite this method - we should get all accounts by **USER phoneNumber**

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        return accountMapper.toDto(accountService.getByAccountNumber(accountNumber));
+    }
+
+    @PostMapping("/transfer")
+    public String transfer(@RequestParam TransactionRequestDto requestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/AccountController.java]: why RequestParam?

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    @GetMapping("/{accountNumber}")
+    public BigDecimal getBalance(@PathVariable String accountNumber) {
+        return accountService.getByAccountNumber(accountNumber).getBalance();
+    }
+
+    @GetMapping("/history/{accountNumber}")
+    public List<TransactionResponseDto> getHistory(
+            @PathVariable String accountNumber,
+            @RequestParam(defaultValue = "1") Integer page,
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/AccountController.java]:
```suggestion
            @RequestParam(defaultValue = "0") Integer page,
```
pages start with 0

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                .stream()
+                .map(transactionMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @PatchMapping("/{accountNumber}")
+    public void block(@PathVariable String accountNumber) {
+        Account account = accountService.getByAccountNumber(accountNumber);
+        account.setActive(false);
+        accountService.save(account);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/AccountController.java:78]: good job!

File: src/main/java/com/example/bank/service/demo/controller/UserController.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+@RequestMapping("/users")
+public class UserController {
+    private final UserService userService;
+    private final UserMapper userMapper;
+
+    @PostMapping
+    public UserResponseDto create(@RequestBody UserRequestDto userRequestDto) {
+        User user = userMapper.toEntity(userRequestDto);
+        Role role = new Role();
+        role.setRoleName(USER);
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/UserController.java]: the roles should be already injected to the db by some `InjectDataComponent` in `@PostCostruct` method, let's just get USER role from the db

File: src/main/java/com/example/bank/service/demo/controller/UserController.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        user.getRoles().add(role);
+        return userMapper.toDto(userService.save(user));
+    }
+
+    @PutMapping("/{userId}")
+    public UserResponseDto update(@RequestBody UserRequestDto userRequestDto,
+                                  @PathVariable Long userId) {
+        User user = userMapper.toEntity(userRequestDto);
+        user.setId(userId);
+        user.setRoles(userService.getById(userId).getRoles());
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/UserController.java:40]: good job!

File: src/main/java/com/example/bank/service/demo/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        return userMapper.toDto(userService.update(user));
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/UserController.java]: we can use save() as well here

File: src/main/java/com/example/bank/service/demo/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    @GetMapping("/by-phone")
+    public UserResponseDto getByPhoneNumber(@RequestParam String phoneNumber) {
+        return userMapper.toDto(userService.getByPhoneNumber(phoneNumber));
+    }
+
+    @DeleteMapping("/{userId}")
+    public String delete(@PathVariable Long userId) {
+        userService.delete(userService.getById(userId));
+        return "User with id: " + userId + " was deleted!";
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/controller/UserController.java]: it is ok for testing purposes, but in general it is better to return void in such cases

File: src/main/java/com/example/bank/service/demo/entity/dto/AccountResponseDto.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.example.bank.service.demo.entity.Currency;
+import java.math.BigDecimal;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class AccountResponseDto {
+    private Long id;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/AccountResponseDto.java:10]: let's add userId as well

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.example.bank.service.demo.entity.dto;
+
+import com.example.bank.service.demo.entity.Account;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class TransactionRequestDto {
+    private Account fromAccount;
+    private Account toAccount;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/TransactionRequestDto.java]: how can we pass entity in requestDto?

please pass ids instead - never pass entities in RequestDto or return them in ResponseDto

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+import java.time.LocalDateTime;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public class TransactionResponseDto {
+    private Long id;
+    private Account fromAccount;
+    private Account toAccount;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java]: same here - return accountNumbers as String, or ids

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    private LocalDateTime dateTime;
+    private BigDecimal amount;
+    private Transaction.Type type;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java]: probably it is better to return type as String

File: src/main/java/com/example/bank/service/demo/entity/dto/UserRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
+package com.example.bank.service.demo.entity.dto;
+
+import java.time.LocalDate;
+import lombok.Data;
+
+@Data
+public class UserRequestDto {
+    private Long id;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/UserRequestDto.java]: where will we get the id?

File: src/main/java/com/example/bank/service/demo/entity/dto/UserResponseDto.java [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Data;
+
+@Data
+@Builder
+public class UserResponseDto {
+    private Long id;
+    private String name;
+    private LocalDate dateOfBirth;
+    private String phoneNumber;
+    private Set<Role> roles;
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/entity/dto/UserResponseDto.java]: don't return entities in dtos

File: src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        return account;
+    }
+
+    @Override
+    public AccountResponseDto toDto(Account account) {
+        return AccountResponseDto.builder()
+                .accountNumber(account.getAccountNumber())
+                .currency(account.getCurrency())
+                .balance(account.getBalance())
+                .build();
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java:37]: not all fields are set

File: src/main/java/com/example/bank/service/demo/service/impl/UserMapperImpl.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Override
+    public UserResponseDto toDto(User user) {
+        return UserResponseDto.builder()
+                .id(user.getId())
+                .name(user.getName())
+                .phoneNumber(user.getPhoneNumber())
+                .dateOfBirth(user.getDateOfBirth())
+                .build();
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/service/impl/UserMapperImpl.java:29]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java (hunk 1 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/AccountController.java]: made correction

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java (hunk 2 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/AccountController.java]: changed by @RequestBody

File: src/main/java/com/example/bank/service/demo/controller/AccountController.java (hunk 3 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/AccountController.java]: done

File: src/main/java/com/example/bank/service/demo/entity/dto/UserRequestDto.java (hunk 13 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/UserRequestDto.java]: deleted. id only in ResponceDto

File: src/main/java/com/example/bank/service/demo/controller/UserController.java (hunk 5 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/UserController.java]: Corrected

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java (hunk 12 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java]: Changed to String

File: src/main/java/com/example/bank/service/demo/entity/dto/UserResponseDto.java (hunk 14 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/UserResponseDto.java]: Made correction

File: src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java (hunk 15 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java:37]: Made correction

File: src/main/java/com/example/bank/service/demo/service/impl/UserMapperImpl.java (hunk 16 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/service/impl/UserMapperImpl.java:29]: Made correction

File: src/main/java/com/example/bank/service/demo/controller/UserController.java (hunk 7 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/UserController.java]: done

File: src/main/java/com/example/bank/service/demo/entity/dto/AccountResponseDto.java (hunk 9 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/AccountResponseDto.java:10]: done

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionRequestDto.java (hunk 10 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/TransactionRequestDto.java]: done

File: src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java (hunk 11 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/entity/dto/TransactionResponseDto.java]: done

File: src/main/java/com/example/bank/service/demo/controller/UserController.java (hunk 8 above)
Comment (Sashk52) [reply src/main/java/com/example/bank/service/demo/controller/UserController.java]: changed

File: src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public AccountResponseDto toDto(Account account) {
+        return AccountResponseDto.builder()
+                .id(account.getId())
+                .accountNumber(account.getAccountNumber())
+                .currency(account.getCurrency().toString())
+                .balance(account.getBalance().toString())
+                .isActive(account.isActive())
+                .build();
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/bank/service/demo/service/impl/AccountMapperImpl.java:37]: let's set userId as well

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**